Root FS Builds with Ubuntu 20.04

dack-fe opened the pull request:
Updated Root FS to Ubuntu 20.04 - Closes #15
Automatically adds Frost Edge and the supporting packages

Closes #22

tvannoy committed:
initial commit of rootfs build script

audiologic-vannoy committed:
first working setup

tvannoy committed:
add nodejs and python

Jevitate committed:
Update packages

audiologic-vannoy committed:
Add everything we need for network discovery

In particular, we are using avahi for mDNS, balena-io wifi-connect
for managing the wifi connection process, and we are using a ralink
wifi adapter that needs firmware files to be installed.

Co-authored-by: Jevitate <55113801+Jevitate@users.noreply.github.com>

audiologic-vannoy committed:
more informative output, remove need for user input

audiologic-vannoy committed:
install dotnet runtime for webapp

fe-wickham committed:
Added script for docker to build the rootfs

dack-fe committed:
Updated Root File System to 20.04

dack-fe committed:
Initial Commit of Jenkinsfile

dack-fe committed:
Updated the file extension

dack-fe committed:
Updated the Jenkinsfile

dack-fe committed:
Build on the correct computer...

dack-fe committed:
Added step to make build script executable

dack-fe committed:
Installing Frost Edge in Root FS

dack-fe committed:
Merge branch 'dev' of https://github.com/fpga-open-speech-tools/utils into rootfs

dack-fe committed:
Updated Jenkinsfile to check for changes to the Root FS Folder

dack-fe committed:
Updated to any date

Commit:
Enable QEMU ARM emulation with update-binfmts instead of copying qemu-arm-static

Suddenly, chrooting into the armhf rootfs is not working in WSL 2 (using the Linux filesystem)
or in Docker on Arch Linux. Explicitly enabling ARM emulation lets QEMU automatically choose
to use ARM emulation when executing ARM binaries. This commit makes chrooting works in WSL 2 and Docker.

fe-wickham committed:
Updated for attempts at 20.04 support with docker building it

dack-fe committed:
Updated the Name of the Service and Script

dack-fe committed:
Seperated Bash Script into multiple services

Created Service for Web App and Deployment Manager
Updated the Packaging to Copy, Install, and Uninstall new services

dack-fe committed:
Updated Type, Description

dack-fe committed:
Updated Description

dack-fe committed:
Updated restart because of a oneshot

fe-wickham committed:
Merge branch 'dev' of github.com:fpga-open-speech-tools/utils into rootfs

fe-wickham committed:
Merge branch 'rootfs' of github.com:fpga-open-speech-tools/utils into rootfs

dack-fe committed:
Updated to the correct download link

fe-wickham committed:
Merge branch 'rootfs' of github.com:fpga-open-speech-tools/utils into rootfs

fe-wickham committed:
Updated docker build script to run the default entrypoint

dack-fe committed:
Updating for non-interactive build

dack-fe committed:
Merge branch 'rootfs' of https://github.com/fpga-open-speech-tools/utils into rootfs

dack-fe committed:
Updated the build scripts

dack-fe committed:
Update the location of the web app

dack-fe committed:
Update the unzip location

dack-fe committed:
Updated the unzip process for the web app

dack-fe committed:
Updated to the correct file name

dack-fe committed:
Added dhclient

dack-fe committed:
Updated the dhclient package name

dack-fe committed:
Added less to package list

dack-fe committed:
Updated the package list

dack-fe committed:
Set strict ssl check to false

dack-fe committed:
Added commands to have DNS lookup, pip3 installs

dack-fe committed:
Added requests install

tvannoy commented:
The frost-edge deb does not get downloaded when the rootfs build is not done in Jenkins. Other than that, the docker build works for me on WSL. 

A straight WSL build is not working for me, but I suspect the problem is on my end; apt complains about invalid gpg signatures:
```
Err:1 http://ports.ubuntu.com/ubuntu-ports focal InRelease
  At least one invalid signature was encountered.
```
which causes a bunch of cascading failures when attempting to install packages. 

When installing the dotnet runtime, we get this warning: 
```
dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.
```
Not a big deal, but we could look into using the installer instead of the script. When I set it up the first time, I found the dotnet script easier to use. 


Looks good overall

There are a couple of gpg key errors for some debian apt sources:
```
W: GPG error: http://cdn-fastly.deb.debian.org/debian jessie Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 7638D0442B90D010 NO_PUBKEY CBF8D6FD518E17E1
W: GPG error: http://cdn-fastly.deb.debian.org/debian stretch Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 7638D0442B90D010 NO_PUBKEY 04EE7237B7D453EC NO_PUBKEY EF0F382A1A7B6500
```
I don't know if those sources are required. As best as I could tell, the `cdn-fastly` sources are commented out in our sources list. It's possible that we may not need the debian repos for the ralink firmware, but that's not very important as it seems to work as is. 

Looks good overall

tvannoy commented on rootfs/docker_build.sh:
```diff
@@ -0,0 +1,6 @@
+docker build -t build_rootfs .
+docker run --name build_rootfs --privileged build_rootfs 
+#docker run --name build_rootfs --privileged --entrypoint /bin/bash -it build_rootfs 
```
delete commented out line?

dack-fe replied:
02d2ff7

tvannoy commented on rootfs/setup_rootfs.sh:
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
+
+# TODO: apt keeps complaining about our locale not being set; we should probably set it properly
```
I believe this is outdated now. It looks like locale/timezone are set up properly now.

dack-fe replied:
7a216c9

tvannoy commented:
We should be able to eliminate the need for using dos2unix by creating the following `.gitattributes` file:
```
* text=auto  eol=lf
```
We can use finer-grained control if we want, e.g. only sh files are forced to use lf endings in the repo and in checkout. 

https://www.git-scm.com/docs/gitattributes
https://docs.github.com/en/free-pro-team@latest/github/using-git/configuring-git-to-handle-line-endings

tvannoy reviewed (changes requested):
We should add a short readme about how to execute the bulid process, e.g. run docker_build.sh

dack-fe committed:
Removed commented out line

dack-fe committed:
Removed comment

dack-fe committed:
Added a check for Frost Edge

dack-fe commented:
> The frost-edge deb does not get downloaded when the rootfs build is not done in Jenkins. Other than that, the docker build works for me on WSL.
   - Updated with commit: c38b421 - Now the Build RootFS Script checks for Frost Edge from Jenkins and if it doesn't find it, the script downloads Frost Edge from Frost Release S3 Bucket.
---
> A straight WSL build is not working for me, but I suspect the problem is on my end; apt complains about invalid gpg signatures:
> ```
> Err:1 http://ports.ubuntu.com/ubuntu-ports focal InRelease
>   At least one invalid signature was encountered.
> ```
> 
> which causes a bunch of cascading failures when attempting to install packages.
   - This was never designed for a straight WSL Build due to permissions.
---
> 
> When installing the dotnet runtime, we get this warning:
> 
> ```
> dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
> dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
> dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
> dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.
> ```
> 
> Not a big deal, but we could look into using the installer instead of the script. When I set it up the first time, I found the dotnet script easier to use.
   - Added as Enhancement Issue #20  
---
> There are a couple of gpg key errors for some debian apt sources:
> 
> ```
> W: GPG error: http://cdn-fastly.deb.debian.org/debian jessie Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 7638D0442B90D010 NO_PUBKEY CBF8D6FD518E17E1
> W: GPG error: http://cdn-fastly.deb.debian.org/debian stretch Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 7638D0442B90D010 NO_PUBKEY 04EE7237B7D453EC NO_PUBKEY EF0F382A1A7B6500
> ```
> 
> I don't know if those sources are required. As best as I could tell, the `cdn-fastly` sources are commented out in our sources list. It's possible that we may not need the debian repos for the ralink firmware, but that's not very important as it seems to work as is.
   - Added as Enhancement Issue #21

dack-fe commented:
> We should be able to eliminate the need for using dos2unix by creating the following `.gitattributes` file:
> 
> ```
> * text=auto  eol=lf
> ```
> 
> We can use finer-grained control if we want, e.g. only sh files are forced to use lf endings in the repo and in checkout.
> 
> https://www.git-scm.com/docs/gitattributes
> https://docs.github.com/en/free-pro-team@latest/github/using-git/configuring-git-to-handle-line-endings

Added as Enhancement Issue #22

dack-fe committed:
Initial ReadMe

tvannoy commented:
> This was never designed for a straight WSL Build due to permissions.
Gotcha. I think I had it working on WSL at some point, but I don't remember. I guess I would have likely needed to build it in the WSL filesystem. 

I didn't have a proper Ubuntu machine to test on, but I imagine it still works fine since that's what I originally developed on. Either way, docker handles any OS so it doesn't matter.

tvannoy commented on rootfs/ReadMe.md:
```diff
@@ -0,0 +1,11 @@
+## Building the Root FS with Docker
+1. Install [Docker for Windows](https://docs.docker.com/docker-for-windows/install/)
+2. Open WSL and Navigate to `[FrOST Repos]\utils\rootfs\`
+3. Run `./build_rootfs.sh`
```
I (and the jenkinfiles) run `docker_build.sh`

dack-fe replied:
Updated with 2b9b109

dack-fe committed:
Updated the ReadMe and Scripts to updated the EoL Characters

dack-fe commented:
> > This was never designed for a straight WSL Build due to permissions.
> > Gotcha. I think I had it working on WSL at some point, but I don't remember. I guess I would have likely needed to build it in the WSL filesystem.
> 
> I didn't have a proper Ubuntu machine to test on, but I imagine it still works fine since that's what I originally developed on. Either way, docker handles any OS so it doesn't matter.

Jenkins is currently building on Ubuntu 20.04 with Docker

tvannoy reviewed (approved):
👍 💯

fe-wickham committed:
Introduce end-of-line normalization and updated README with more info on key files

fe-wickham commented:
> A straight WSL build is not working for me, but I suspect the problem is on my end; apt complains about invalid gpg signatures:

Did you try running it directly through WSL on the Windows filesystem or the WSL filesystem? It will fail running it directly on the Windows filesystem because it can't handle the chroot properly. I recall getting errors in that same area but I don't know if they were the exact same error from running it on the Windows filesystem.

Edit:
>  I guess I would have likely needed to build it in the WSL filesystem.

Just saw you posted this, that is almost certainly the issue.

fe-wickham commented:
> We should be able to eliminate the need for using dos2unix by creating the following `.gitattributes` file:
> 
> ```
> * text=auto  eol=lf
> ```
> 
> We can use finer-grained control if we want, e.g. only sh files are forced to use lf endings in the repo and in checkout.
> 
> https://www.git-scm.com/docs/gitattributes
> https://docs.github.com/en/free-pro-team@latest/github/using-git/configuring-git-to-handle-line-endings

I agree and needing to run `dos2unix` becomes rather annoying pretty quickly since things dont just work. I have fixed this with de1b2e9 which closes #22

fe-wickham reviewed (approved):
Looks good

tvannoy commented:
> 
> 
> > A straight WSL build is not working for me, but I suspect the problem is on my end; apt complains about invalid gpg signatures:
> 
> Did you try running it directly through WSL on the Windows filesystem or the WSL filesystem? It will fail running it directly on the Windows filesystem because it can't handle the chroot properly. I recall getting errors in that same area but I don't know if they were the exact same error from running it on the Windows filesystem.
> 
> Edit:
> 
> > I guess I would have likely needed to build it in the WSL filesystem.
> 
> Just saw you posted this, that is almost certainly the issue.

Just built it in the WSL filesystem -- worked flawlessly. I now recall @fe-wickham mentioning chroot issues in the Windows filesystem before... 

I can also confirm that the frost-edge deb download worked, and there was no need to run dos2unix on a fresh clone, so the eol conversions seemed to work as well. 

It's beautiful! 🥳

tvannoy commented:
I did get an npm error about certificates, though -- not sure if it's actually important:
```
npm ERR! code UNABLE_TO_GET_ISSUER_CERT_LOCALLY
npm ERR! errno UNABLE_TO_GET_ISSUER_CERT_LOCALLY
npm ERR! request to https://registry.npmjs.org/assert-plus/-/assert-plus-1.0.0.tgz failed, reason: unable to get local issuer certificate
```

I also get some warnings around sudo when the wifi-connect installer runs: 
```
sudo: unable to resolve host trevvvyx1carbon: Name or service not known
```
I suspect this is because we are running as root, so sudo doesn't really make sense. This isn't really an issue, but we could squelch that warning by using sed to remove `sudo` from the wifi-connect install script.

dack-fe commented:
> I did get an npm error about certificates, though -- not sure if it's actually important:
> 
> ```
> npm ERR! code UNABLE_TO_GET_ISSUER_CERT_LOCALLY
> npm ERR! errno UNABLE_TO_GET_ISSUER_CERT_LOCALLY
> npm ERR! request to https://registry.npmjs.org/assert-plus/-/assert-plus-1.0.0.tgz failed, reason: unable to get local issuer certificate
> ```
 - I am not seeing these NPM Errors in Jenkins. @tvannoy Did you test the Frost Web App?

---
> I also get some warnings around sudo when the wifi-connect installer runs:
> 
> ```
> sudo: unable to resolve host trevvvyx1carbon: Name or service not known
> ```
> 
> I suspect this is because we are running as root, so sudo doesn't really make sense. This isn't really an issue, but we could squelch that warning by using sed to remove `sudo` from the wifi-connect install script.
 - We do not use a sudo call when configuring the WiFi

dack-fe committed:
Removed dos2unix calls

dack-fe committed:
Merge branch 'rootfs' of https://github.com/fpga-open-speech-tools/utils into rootfs

tvannoy commented:
> > I did get an npm error about certificates, though -- not sure if it's actually important:
> > ```
> > npm ERR! code UNABLE_TO_GET_ISSUER_CERT_LOCALLY
> > npm ERR! errno UNABLE_TO_GET_ISSUER_CERT_LOCALLY
> > npm ERR! request to https://registry.npmjs.org/assert-plus/-/assert-plus-1.0.0.tgz failed, reason: unable to get local issuer certificate
> > ```
> 
>  I am not seeing these NPM Errors in Jenkins. 

Interesting, I am getting them in the docker build and the WSL build. It appears to be solved by `npm config set strict-ssl false`, which is in frost-edge packaging scripts; I wonder if the version that is being downloaded from S3 does not have that npm setting. 

> @tvannoy Did you test the Frost Web App?

I am out of town, so I do not currently have an audio mini with me to test with.

> > I also get some warnings around sudo when the wifi-connect installer runs:
>
> We do not use a sudo call when configuring the WiFi

We don't, but [Balena's install script](https://raw.githubusercontent.com/balena-io/wifi-connect/master/scripts/raspbian-install.sh) that we download does.

dack-fe commented:
Updated the Frost Edge Package on S3.

fe-wickham committed:
Testing build with NetworkManager fixes

fe-wickham committed:
Added missing rootfs files

fe-wickham committed:
Updated rootfs build to cp Network files into the rootfs not just from the rootfs to somewhere else in it

fe-wickham committed:
Make directory to path before copying files

fe-wickham committed:
Moved Network Manager configuration until post install of NetworkManager

dack-fe committed:
Updated the name of the tar.gz artifcat

dack-fe committed:
Updated the final file

fe-wickham committed:
Install and configure netplan to ensure dhcp runs correctly at boot

fe-wickham committed:
Fixed inconsistent spacing and moved avahi and NetworkManager config to be copied in after setup

fe-wickham committed:
Fixed netplan config
